drivers: can: mcp2515: Add driver for MCP2515 CAN controller

karstenkoenig opened the pull request:
The MCP2515 is a CAN controller that can be connected via SPI to an
host MCU. This driver adds support for the MCP2515 as a new driver in
the CAN subsystem.
As it is a SPI peripheral it uses a thread for its interrupt
handling and the received message filtering is done inside this
interrupt thread, as the MCP2515 filter capabilities are not sufficient
for the Zephyr CAN interface.
The driver was validated with an external CAN logger and the adjusted
CAN sample application.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig commented:
@alexanderwachter 
Heihei, please also review this. I know you have plans for changing the CAN interface further, and I am open for discussions and changes.

Some notes on this driver:
* filtering is done on the host mcu, the mcp2515 doesn't have filtering that matches the CAN rx interface and is generally a lot weaker
* dts not straightforward, reintroduced timing configuration in KConfig.mcp2515 for now
* currently no error feedback, messages will be retried forever until error condition is cleared

alexanderwachter commented:
@karstenkoenig The timing should be in in the device-tree to have the ability to set different speeds on more than one controller. Moreover when you have various boards with different clock speeds its annoying to change the Kconfig for every board.
The error feedback is a little bit tricky. Maybe we could use the timeout for the int_sem too and delete the message from the buffer if a timeout is triggered. But for now the timeout is only for waiting for an empty message box.

karstenkoenig commented:
Having CAN timing information in device tree files is straightforward for on MCU devices, as you can just toss the information into the existing board files, but right now for 'external peripherals' it is harder to provide a straight forward sample until dts overlays/shields are enabled. Most sensors etc all still have that sort of information in KConfig files
https://github.com/zephyrproject-rtos/zephyr/pull/6837

I am not really sure under which circumstance the stm will feedback tx errors now, won't it also try forever? Or will it pretty much report the error and still have transmitted the message? 

The XOR filter idea is really nice, but to configure the filters on the MCP2515 it needs to be put out of operation into configuration mode, so everytime the filters get changed it needs to be recalculated and applied while stopping actual operation, so it might also create unexpected behavior.

alexanderwachter commented:
Now I get your point with the device-tree. I think it's ok for the external controller to have this in Kconfig.

There is no error feedback. The controller will try sending forever as long as there is no ACK . I think that is the behaviour recommended by the ISO standard. The API for reading the bus states (Error active, Error passive and bus off) is still pending. Also recovery from bus off is not possible with the current API.
 
Could you use the Acceptance Filter as they are and just "or" the masking together? In my opinion it is sufficient ho have only six filter banks. The stm32 driver stops operation of the filters too for the short period of changing them. Not using the hardware filtering at all will result in a heavy CPU load depending on the overall bus load.

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+#define MCP2515_CANINTF_ERRIF           BIT(5)
+#define MCP2515_CANINTF_WAKIF           BIT(6)
+#define MCP2515_CANINTF_MERRF           BIT(7)
+
```
form line 16 to here, you could move that to a dedicated header file.

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+	return spi_transceive(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg,
+			      &tx, &rx);
+}
+
```
About all the above spi related function: would there be a way to make it in one function? (just asking, and if you prefer that way, just keep it.)

karstenkoenig replied:
I thought about it, but the methods are quite different and I don't want to throw distinguishing it to the callers. I renamed them with _cmd_ to indicate they are doing remote calls on the mcp2515.

tbursztyka replied:
ok

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+{
+	u8_t opcode_buf[2];
+
+	opcode_buf[0] = MCP2515_OPCODE_WRITE;
```
rewrite that: 

u8_t opcode_buf = { ..., ... };

Apply that comment everywhere. (you did it properly in soft-reset function above actually)

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+{
+	u8_t canstat;
+
+	mcp2515_bit_modify(dev, MCP2515_ADDR_CANCTRL, 0x07 << 5, mode << 5);
```
0x07, 5: could you create macro to these with relevant names?

karstenkoenig replied:
I tried to do my best, there is currently 3 functions left where it is still using unnamed bit positions and masks
mcp2515_convert_can_msg_to_mcp2515_frame
mcp2515_convert_mcp2515_frame_to_can_msg
mcp2515_configure

They won't really become easier to read nor to follow, as you need to look at the datasheet if you want to know the details, and then it is easier to check for correctness when having the bare positions spilled out instead of double checking defines.
I can change the rest too if it is policy or just frowned upon.

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+		target[MCP2515_FRAME_OFFSET_EID0] = source->ext_id;
+	}
+
+	u8_t rtr = 0;
```
Never declare a variable in the middle of a code block, always a the beginning  of it. Apply that everywhere.

karstenkoenig replied:
Done

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+		target->std_id = (source[MCP2515_FRAME_OFFSET_SIDH] << 3) |
+				 (source[MCP2515_FRAME_OFFSET_SIDL] >> 5);
+	}
+	target->dlc = source[MCP2515_FRAME_OFFSET_DLC] & 0x0F;
```
add an empty line after '}' unless it's another } etc... Apply that everywhere.

karstenkoenig replied:
Done

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+	k_mutex_lock(&dev_data->filter_mutex, K_FOREVER);
+
+	for (; filter_idx < CONFIG_CAN_MAX_FILTER; filter_idx++) {
+		if (BIT_LL(filter_idx) & dev_data->filter_usage) {
```
too many if imbrications.

I guess you could reverse that test and use continue, to reduce indentation of the lines below.
If necessary, it's also a good idea to create an inline function to de-clutter the code.

karstenkoenig replied:
Done

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,748 @@
+			K_PRIO_COOP(dev_cfg->int_thread_priority), 0, K_NO_WAIT);
+
+	/* Initialize TX data */
+	k_sem_init(&dev_data->tx_sem, 3, 3);
```
For all of these sem/mutex maybe you could try using K_SEM_INITIALIZER()/K_MUTEX_INITIALIZER(). It will generate a little less code then. (and it's faster to initialize).

karstenkoenig replied:
These macros are flagged internal, does that mean I can use them in kernel drivers or is that just for really core components? Use throughout the rest of the tree seems to lean towards k_sem_init/k_mutex_init style initialization.

I more dislike the general amount of mutexes and semaphores, but couldn't come up with a smaller amount

tbursztyka replied:
It's ok to use these macros in native drivers, not in application.

karstenkoenig replied:
Ok, I changed the initializers to be in the mcp2515_data_1 initializer, thanks!

tbursztyka commented on samples/can/prj.conf.mcp2515:
```diff
@@ -0,0 +1,19 @@
+# CAN + MCP2515 config
```
you could split that into another patch.

karstenkoenig replied:
Done

karstenkoenig replied:
Done

karstenkoenig committed:
drivers: can: mcp2515: Add driver for MCP2515 CAN controller

The MCP2515 is a CAN controller that can be connected via SPI to an
host MCU. This driver adds support for the MCP2515 as a new driver in
the CAN subsystem.
As it is a SPI peripheral it uses a thread for its interrupt
handling and the received message filtering is done inside this
interrupt thread, as the MCP2515 filter capabilities are not sufficient
for the Zephyr CAN interface.
The driver was validated with an external CAN logger and the adjusted
CAN sample application.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig commented:
Thanks for the great feedback so far! I tried incorporating everything from @tbursztyka and updated the pull request.
I will work on the filtering in another commit, @alexanderwachter your idea has some caveats such as filter 2-5 will overflow easier, but in general it sounds like a very sensible approach that should at least stem the message tide

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,618 @@
+
+	mcp2515_convert_can_msg_to_mcp2515_frame(msg, tx_frame);
+	mcp2515_cmd_write_reg(dev, addr_tx_ctrl + MCP2515_ADDR_OFFSET_CTRL2FRAME,
+			tx_frame, MCP2515_FRAME_LEN);
```
tiny alignement issue (should be below last '(' + 1 space)

apply that rule everywhere (I probably did not notice it the first time)

karstenkoenig replied:
Sorry I have to double check here, as the first time I did it checkpatch complained about space instead of tabs
You want any link breaks aligned under and after the last current open bracket right? It doesn't have to be exactly 1 behind the last open bracket, right? Anyways, I did just that and checkpatch didn't complain (but keeps seeing lines longer than 80 chars that aren't, does it tab extend to 8 space?)

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,618 @@
+	.tx_mutex = _K_MUTEX_INITIALIZER(mcp2515_data_1.tx_mutex),
+	.tx_sem = _K_SEM_INITIALIZER(mcp2515_data_1.tx_sem, 3, 3),
+	.tx_cb = {
+			{.cb = NULL, .sem = _K_SEM_INITIALIZER(mcp2515_data_1.tx_cb[0].sem, 0, 1) },
```
does that fit the 80chars limit?

tbursztyka reviewed (commented):
just some minor comments, rest is fine to me.

alexanderwachter commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,618 @@
+static int mcp2515_cmd_bit_modify(struct device *dev, u8_t reg_addr, u8_t mask,
+				  u8_t data)
+{
+	u8_t cmd_buf[4] = { MCP2515_OPCODE_BIT_MODIFY, reg_addr, mask, data };
```
Could you write
u8_t cmd_buf[] = { MCP2515_OPCODE_BIT_MODIFY, reg_addr, mask, data };

alexanderwachter commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,618 @@
+	u8_t cmd_buf[4] = { MCP2515_OPCODE_BIT_MODIFY, reg_addr, mask, data };
+
+	const struct spi_buf tx_buf = {
+		.buf = cmd_buf, .len = 4,
```
and
.buf = cmd_buf, .len = sizeof (cmd_buf),

alexanderwachter commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,618 @@
+	/* will enter configuration mode automatically */
+	mcp2515_cmd_soft_reset(dev);
+
+	mcp2515_cmd_write_reg(dev, MCP2515_ADDR_CNF3, config_buf, 4);
```
I would prefer sizeof (config_buf) instead of just 4

alexanderwachter reviewed (changes requested):
Looks good. Some minor comments.

alexanderwachter commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,619 @@
+		{ .buf = NULL, .len = buf_len }
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf, .count = sizeof(tx_buf)/sizeof(tx_buf[0])
```
You could use ARRAY_SIZE(tx_buf) here

karstenkoenig commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -0,0 +1,132 @@
+	range 1 8
+	help
+	  Time quantums of phase buffer 1 segment (ISO 11898-1)
+
```
I wanted to avoid doing any unexpected math, and there is hard limits on the values, but I think you are right, I could just do the following to allow for full range and stay in the bounds, will keep it in mind when switching to dts proper
tq_prop = tq_prop_bs1 / 2
tq_bs = (tq_prop_bs1 / 2) + (tq_prop_bs1 % 2)

alexanderwachter replied:
I had
tq_bs1 = (tq_prop_bs1 -1) & 0x07
tq_prop = tq_prop_bs1 - tq_bs1 - 2
in mind but its the same :-)

karstenkoenig commented:
Thanks for all the reviews and sorry for silently pushing, took it a bit slow. 

@alexanderwachter 
I thought a bit more about the filter suggestions and would want to discuss a little more. My background is ebikes, and there isn't much CAN data flying around, so the filtering isn't really that critical and you can better judge the situation

As there is only two filter masks on the mcp2515 the actual filtering will be very limited. My worry with random or'ing the masks together will be unintended filters such as worst cases ending up with 0 masks that won't filter. What do you think would be a decent strategy that allows application developers to pick sensible filters and leverage the limited filtering the best way using the current can interface?

1. straight forward would be allow only two filters overall, that forces good filter decisions in the first place by the application developer and will make him/her aware to pitfalls
2. allow 32 filters but create a filter each for standard and one for extended frames on mcp2515, or the masks, xor the filters, try validating them such as checking for 0 masks, run final filter check on host mcu
3. allow max 6 filters similar to 2., but messy because first inbox of mcp2515 has 1 mask and two filters and the second inbox has 1 mask and 4 filters, especially confusing with standard and extended frames, final filter check on host mcu

I prefer either 1. or 2., as they are the easiest to explain what is happening and don't try to paper over limitations.
2. could lead to unintended results but is more powerful than 1. if different paths for messages is important

Should I open a discussion issue ticket for this? Might be a better place then drag it out in a pull request

alexanderwachter commented:
@karstenkoenig I have to study the data sheet an think about the filters.
I will comment it next week

alexanderwachter commented:
@karstenkoenig 
2.) sounds good but seems to be complicated. 
1.)  is really restricted and does not utilize all the mcp2515 capabilities.
Actually I don't know what's better.

Please have a look at https://github.com/alexanderwachter/zephyr/tree/can_extend_api
This is the API extension for bus-state-handling. Do you agree on it?
Unfortunately I have no idea how to test it because I dont't know how to produce error frames on purpose.

karstenkoenig commented:
@alexanderwachter 
On the filters:
I checked with the messages I am usually dealing with and there is literally no way to sensibly put together one or two filters, the message id's are just so different there is barely a common mask possible, even looking at just a handful of messages. 
To be honest I am prefering 1) now for the most simple use cases where people must be clever to pick their filters. Actually I think all those queues or callbacks for the filters are pretty massive and it seems easier to me to just dump it all into a single rx queue or callback that can dispatch messages by higher level logic. So not attaching a queue or callback to a filter. Why did you prefer to have a queue/callback per filter? 

On the bus error conditions:
I didn't have to deal with bus errors so far

The get_state I'd rather name get_error_state. Right now there is no 'disable CAN device' function but there might be a future and that is properly the more relevant 'state'.

Also do you specifically need the recover method? Couldn't you just configure it to auto-recover? I'd rather not expose device features until there is a use case for it, because that will make sure it at least gets used and tested and I can't see why I wouldn't want my CAN controller to auto recover for the most time.

alexanderwachter commented:
@karstenkoenig I think 1) should the preferred solution or a Kconfig option to switch from your actual solution to 1) or just leave it as it is now.
I preferred one queue/callback per filter because the stm32 (and maybe NXP LPC series) have those hardware filtering capabilities. This helps reducing the interrupt flood when having high bus loads. I thought it's a "nice to have" when the dispatcher is hidden in the driver. Usually you need this kind of message dispatching anyway.

On the bus-state API:
In my opinion auto recovery is not always a good idea. A node is in bus-off state because it was disturbing the other nodes with sending error-frames and sending erroneous messages. An erroneous node could in a worst case block important frames from being sent.

I will change get_state to get_bus_state.

DWagenk commented:
Hi,
Any chance of getting this PR back to live? 
As far as I can tell not much has changed in the CAN subsystem, so it should be basically working. I've got a test setup here, using zephyr 1.13 release and sligtly modified commits from this PR, that can run the CAN Sample.

I'd be willing to help, at least with testing.

alexanderwachter commented:
Sure. 
@laperie @carlescufi @jukkar @tbursztyka What do you think?

karstenkoenig commented:
Hi, sorry for not pursuing this to completion back in the day. Got tied up with life. I rebased on current master, switched to the new logger and adjusted for some MISRA-C stuff that was added to the stm32_can.c driver as well.
The sample project compiles and it should still be working. But I can not test it for the next months as my development boards are all in some boxes where I don't have access to. @DWagenk I presume my changes are similar to what you did to get it running again, can try running the current state again please?

About the open comments to improving the filter setup, to be honest I wouldn't bother unless someone comes up with a good way to do it and some validation use case. The capabilities of the controller are just not there and it would add complexity without proven gain.

karstenkoenig commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,620 @@
+#include <spi.h>
+#include <gpio.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_CAN_LEVEL
```
Done, thanks

alexanderwachter commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,621 @@
+}
+
+static int mcp2515_cmd_bit_modify(struct device *dev, u8_t reg_addr, u8_t mask,
+									u8_t data)
```
Looks like your editor uses 4 spaces tabs. Zephyr and Linux use 8 spaces tabs.

karstenkoenig replied:
I ran uncrustify again and it seems to have realigned them properly to 4 spaces tab

karstenkoenig replied:
Arrr 8 spaces tab I mean :-)

alexanderwachter commented:
@karstenkoenig I also think we can leave the filters as they are. It's better to mimic a actual device than dealing with the limitations in userspace. As you told me there is a successor to MCP2515 (MCP2517FD) that could deal with the filter requirements.

karstenkoenig commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,621 @@
+	.int_thread_stack = mcp2515_int_thread_stack,
+	.int_sem = _K_SEM_INITIALIZER(mcp2515_data_1.int_sem, 0, UINT_MAX),
+
+	.tx_mutex = _K_MUTEX_INITIALIZER(mcp2515_data_1.tx_mutex),
```
Reworked it back to k_mutex_init and k_sem_init, was no big deal

DWagenk commented:
@karstenkoenig I'll take a closer look by the end of this week and also test the changes on hardware.

carlescufi commented:
@tbursztyka can you please take another look?

tbursztyka commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,621 @@
+}
+
+static int mcp2515_cmd_write_reg(struct device *dev, u8_t reg_addr,
+									u8_t *buf_data, u8_t buf_len)
```
What has happened here with indentation?

It's messed up in numerous places in this file.

karstenkoenig replied:
Looks like you looked at an outdated version, probably what @alexanderwachter spotted and what was fixed by uncrustify, the alignments for the function headers look correct to me with 8 spaces tab.

karstenkoenig replied:
Sorry I was wrong about this, you looked at the commit adding the driver and not the full pull request. The commit with the driver was still faulty, the indentions were fixed in the sample commit, which was obviously wrong. 
Anyways, fixed the 80 char issues and also the indention and it's all properly in the first commit now that adds the driver

tbursztyka reviewed (changes requested):
Looks like you have 80chars limit issue

jfischer-phytec-iot commented:
@karstenkoenig Can you please fix style issues.

I suggest to add it to 1.14 in order to have more diversity with CAN drivers.

galak reviewed (changes requested):
This needs updating to use SPI for dts.  All spi client drivers now utilize DTS.

karstenkoenig committed:
drivers: can: mcp2515: Rework for DTS SPI bindings

Adjusted the MCP2515 driver to switch from KConfig SPI configuration to
DTS based configuration.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig committed:
samples: can: Added project configuration for MCP2515

Project configuration to run the CAN sample with the MCP2515 attached
via SPI as the CAN controller.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig commented:
Fixed the 80char issue.
I added the DTS changes based on this commit for a similar SPI based peripheral controller 
https://github.com/zephyrproject-rtos/zephyr/commit/4118b8843f3c6e6500643b3b5eff70675b3afd68#diff-008346ee80cb97a372b2853a3cc20107

Also adjusted the sample accordingly. Right now I left the DTS change as a separate commit, I can merge these if needed.

@DWagenk 
I don't know how much you used of the sample, but the new sample should also cover how to use the driver with SPI via DTS. Again I can not test this for the near future :-(

jukkar commented on samples/drivers/CAN/prj.conf.mcp2515:
```diff
@@ -0,0 +1,6 @@
+# CAN + MCP2515 config
```
We should make prj.conf generic and use samples/drivers/CAN/boards directory for any board specific config. Probably not worth fixing in this PR (unless you really want to) but at least to be done later.

DWagenk commented:
I finally got to trying the updated PR on Hardware. I used a a custom stm32_min_dev based board with an mcp2515 on it. Tested in Loopback and non-loopback mode (with modified CAN-Sample main.c, that initializes CAN_1 in CAN_NORMAL_MODE).
I fixed some problems in the  SPI with CS via GPIO configuration and now it works. SPI using the STM32 SPI drivers chip-select behaviour didn't work for me, but that could also be a misconfiguration on my part.

Theres a fixup commit to show the necessary changes here https://github.com/DWagenk/zephyr/commit/4a759f5c94407d223548a4a75b7b1435a1eb8e9f

I also changed the CAN sample to use the LED and button provided by the boards dts file. :qatar: See the commit here https://github.com/DWagenk/zephyr/commit/0df72b598acca083d90076ad1f3402116164b24a

@karstenkoenig would you just integrate those changes? I didn't sign them off or check the formatting, because I think they should be integrated into your commits. Just prepared them as seperate commits, to show the fixups I needed for running the sample.

karstenkoenig commented:
@DWagenk Awesome, thanks!
I merged your DTS fix into the  commit that changes the driver to use DTS.

I didn't pick the samples change. It is the right thing to do, but it is not directly related to the driver, also @jukkar suggested further changes by using some board files for in the samples directory.
Or is everybody else ok to just add this here as well?

galak commented on dts/bindings/can/microchip,mcp2515.yaml:
```diff
@@ -0,0 +1,30 @@
+      type: compound
+      category: required
+      generation: define, use-prop-name
+    cs-gpios:
```
cs-gpios shouldn't be here, its handled at the spi controller level.

karstenkoenig replied:
Hmmm I just added this because @DWagenk couldn't get the sample to work otherwise. 
https://github.com/DWagenk/zephyr/commit/4a759f5c94407d223548a4a75b7b1435a1eb8e9f

Does it have to be DT_MICROCHIP_MCP2515_0_CS_GPIOS_PIN or DT_MICROCHIP_MCP2515_0_CS_GPIO_PIN ? Because https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ethernet/eth_enc28j60.c is doing it in singular (GPIO instead of GPIOS) According to the spi.yaml it should be GPIOS as far as I understand

DWagenk replied:
@galak I looked into the SPI-drivers sources a little bit. Am I understanding correctly, that all of spi-cs via gpio handling is handled at that level already and there's no need to have a spi-cs-gpio appear anywhere in the mcp2515 driver?
So deleting `cs-gpios` section here again and also removing the `config CAN_MCP2515_GPIO_SPI_CS` option from Kconfig.mcp2515 as well as the `#ifdef CONFIG_CAN_MCP2515_GPIO_SPI_CS` sections from mcp2515.c would be the way to go?
This would also apply to the enc28j60 that @karstenkoenig used as a reference for his implementation, correct?

karstenkoenig replied:
I removed the cs-gpios section in the mcp2515 dts bindings, as the one in spi.yaml should be enough. 
As the enc28j60 driver is still using the CONFIG_*_GPIO_SPI_CS it is left in for now here as well. I can probably #ifdef DT_MICROCHIP_MCP2515_0_CS_GPIOS_PIN as it is an optional field, but without any clear indication which would be the right way I'd rather have it consistent at least :-|

galak reviewed (dismissed):
One DTS change needed.

ulfalizer commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -0,0 +1,141 @@
+
+config CAN_MCP2515
+	bool "MCP2515 CAN Driver"
+	default n
```
`n` is the implicit for `bool` symbols, so can skip the `default`. That makes it consistent, and generates clearer docs ("Implicitly defaults to n" instead of " - n if <bunch of inherited dependencies>").

There's a page with some [Kconfig tips and best practices](https://docs.zephyrproject.org/latest/tools/kconfig/index.html) by the way.

karstenkoenig replied:
All fixed, thanks

ulfalizer commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -0,0 +1,141 @@
+if CAN_MCP2515_GPIO_SPI_CS
+config CAN_MCP2515_SPI_CS_PORT_NAME
+	string "GPIO driver's name to use to drive SPI CS through"
+	default ""
```
Can skip this default too. `string` symbols default to the empty string (and you get clearer docs).

ulfalizer commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -0,0 +1,141 @@
+	int "Oscillator frequency"
+	default 8000000
+	help
+	 Specify the frequency of the oscillator connected to the MCP2515.
```
Nitty, but help texts are usually indented by two spaces instead of one.

karstenkoenig replied:
Fixed, must have slipped :-)

ulfalizer reviewed (changes requested):
Just some nits.

ulfalizer reviewed (approved):
Nits fixed.

karstenkoenig commented:
@DWagenk apparently @alexanderwachter wrote some api test in tests/drivers/can/api, can

alexanderwachter commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,639 @@
+	if (gpio_pin_enable_callback(dev_data->int_gpio, dev_cfg->int_pin)) {
+		return -EINVAL;
+	}
+
```
Can you add a bitrate option to Kconfig and add call mcp2515_configure here? The device should be initialized and ready to use after init.

karstenkoenig replied:
Done

DWagenk commented:
Finally got to try out the current state of this PR, but didn't get it to work yet. It fails in `mcp2515_set_mode` and returns `-EIO`.
By what I see that could be just an SPI problem in my setup (it's the first read of MCP2515 registers), so I'll have to look into that before I can give more feedback on this PR.

karstenkoenig committed:
drivers: can: mcp2515: Fixes for DT SPI cs

Fixed using chipselect with seperate chipselect GPIOs and how they were
referenced from/in DeviceTree.
Also configure the device during initialization so it's ready to go
after init.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig commented:
I have access to my stuff again and fixed the driver. 

@DWagenk you are right and probably ran into SPI chipselect issues, I made a mistake with the devicetree connection. Fixed it and provided an overlay dts file in the sample project for reference. To be honest the interworking of different parts of devicetree/kconfig etc are kind of hard to follow.... the cs-gpios dts struct has to be part of the spi bus and not the bus member, and when mapped to defines in the build process they lose the plural s and end up being CS_GPIO_xyz etc :-( 

Anyways, I also ran the CAN api testsuite and it passes

DWagenk commented:
@karstenkoenig great that you picked it back up and have your Hardware for testing. I'll keep trying it out and hope it helps if I just document my issues. Iwill comment directly on the files causing issues for me.

DWagenk commented on samples/drivers/CAN/mcp2515-dts.overlay:
```diff
@@ -0,0 +1,24 @@
+	};
+};
+
+&can1 {
```
can1 seems to only exists on boards, that already have other can hardware. For me it compiles on
 - stm32f072b_disco

but not on 
 - nucleo_f091rc
 - stm32_min_dev

Also the generated fixup `DT_CAN_1_BUS_SPEED` on stm32f072b_disco board points at the stm32 CAN driver.

DWagenk commented on samples/drivers/CAN/mcp2515-dts.overlay:
```diff
@@ -0,0 +1,24 @@
+&spi1 {
+	status = "ok";
+	cs-gpios = <&gpioa 4 GPIO_DIR_OUT>;
+	mcp2515@0 {
```
tried fixing the issue described below by changing this to 
`can1: mcp2515@0 {`
Creates issue of duplicate label for stm32f072b_disco board, lets cmake run successfully for the other boards, but has issues with  fixup `DT_CAN_1_BUS_SPEED` missing during compile step for nucleof091rc and stm32_min_dev board

karstenkoenig commented:
DeviceTree is so beautiful and so utter frustrating at the same time...
I moved the bus-speed in the sample into the overlay dts file, and am now using the MCP2515_0_BUS_SPEED in the mcp2515.c driver. This works and does not need an existing can1 section at all in the dts.
To double check I just hooked it up and ran it on an nrf52_dk10040 board, the only thing that needed changing were the GPIOs in the sample config

DWagenk commented:
@karstenkoenig that fixed it! It compiles and runs the CAN sample in loopback mode now (stm32_min_dev board with external button).

I'd propose merging it now and then doing little improvements later.

LGTM

DWagenk commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -0,0 +1,80 @@
+
+if CAN_MCP2515
+
+config CAN_MCP2515_OSC_FREQ
```
I guess these could/should go into device tree as well?
- CAN_MCP2515_OSC_FREQ
- CAN_PROP_SEG
- CAN_PHASE_SEG1
- CAN_PHASE_SEG2
- CAN_SJW

karstenkoenig replied:
Hm yes, the segment length ones are sort of defined already in the can devicetree, but seg1 and prop are merged together there but the mcp2515 allows specifing them independently. I'll try sorting out again why I didn't just use it before

alexanderwachter replied:
@karstenkoenig it should be save to merge CAN_PROP_SEG and CAN_PHASE_SEG1. Even the CAN standard says so. I have no idea why this is separate in the mcp2515. They have the same effect.

DWagenk commented on drivers/can/mcp2515.c:
```diff
@@ -0,0 +1,651 @@
+	/* CNF3, CNF2, CNF1, CANINTE */
+	u8_t config_buf[4];
+
+	if (bitrate == 0) {
```
Maybe we should make it more explicit somewhere, that bus-speed confgirued in device-tree will be ignored, if can_configure is called with `bitrate != 0` ?

alexanderwachter replied:
Indeed. This should be added in the API documentation.

karstenkoenig replied:
haha actually I only found this because the api test tripped there and added it in after double checking with the stm32 can drivers :-)

DWagenk commented on samples/drivers/CAN/prj.conf.mcp2515:
```diff
@@ -0,0 +1,13 @@
+CONFIG_CAN_1=y
+CONFIG_CAN_STM32=n
+CONFIG_CAN_MCP2515=y
+CONFIG_GPIO_LED_DEV="GPIOB"
```
already discussed earlier but noted again here as a to do for later:
use `led0` and `sw0` from the board's device tree.

DWagenk commented on samples/drivers/CAN/mcp2515-dts.overlay:
```diff
@@ -6,13 +6,15 @@
 	status = "ok";
+	cs-gpios = <&gpioa 4 GPIO_DIR_OUT>;
 	mcp2515@0 {
 		compatible = "microchip,mcp2515";
```
If we give it a label properties can be changed in additional (e.g. application specific) overlay files.

```dts
&spi1 {
	/* ... */
        mcp2515_can1: mcp2515@0 {
		/* ... */
	};
};

/* in different file, overlay or position: */
&mcp2515_can1 {
	bus-speed = <250000>;
};
```

dwagenk reviewed (approved):
suggested changes for a follow-up PR

karstenkoenig commented:
@DWagenk Nice! Glad to hear it works for you as well!

alexanderwachter commented:
@karstenkoenig @DWagenk next big change is #13262
